tensorboard/plugins/hparams/list_session_groups.py

```diff
@@ -0,0 +1,411 @@
+    if infos_tuple.end_info is not None:
+      result.status = infos_tuple.end_info.status
+      result.end_time_secs = infos_tuple.end_info.end_time_secs
+    return result
+
+  def _build_session_metric_values(self, session_name):
+    # result is a list of api_pb2.MetricValue instances.
+    result = []
+    metric_infos = self._context.experiment().metric_infos
+    for metric_info in metric_infos:
+      metric_name = metric_info.name
+      try:
+        metric_eval = metrics.last_metric_eval(
+            self._context.multiplexer(),
+            session_name,
+            metric_name)
+      except KeyError:
+        # It's ok if we don't find the metric in the session.
+        # We skip it here. For filtering and sorting purposes its value is None.
+        continue
+
+      # metric_eval is a 3-tuple of the form [wall_time, step, value]
+      result.append(api_pb2.MetricValue(name=metric_name,
+                                        wall_time_secs=metric_eval[0],
+                                        training_step=metric_eval[1],
+                                        value=metric_eval[2]))
+    return result
+
+  def _build_group_from_sessions(self, sessions, session_infos_by_name):
+    assert sessions  # Make sure sessions is non-empty
+    # Sort sessions by name so the order is deterministic.
+    sessions = sorted(sessions, key=lambda session: session.name)
+    # TODO(erez): Do proper metric aggregation. For now we just take
+    # the metric values from the first session.
+    result = api_pb2.SessionGroup(
+        name=session_infos_by_name[sessions[0].name].start_info.group_name,
+        metric_values=sessions[0].metric_values,
+        sessions=sessions,
+        monitor_url=(
+            session_infos_by_name[sessions[0].name].start_info.monitor_url
+        ),
+    )
+    # Copy hparams from the first session (all sessions should have the same
+    # hyperparameter values) into result.
+    # There doesn't seem to be a way to initialize a protobuffer map in the
+    # constructor.
+    hparams = session_infos_by_name[sessions[0].name].start_info.hparams
+    for (key, value) in six.iteritems(hparams):
+      result.hparams[key].CopyFrom(value)
+    return result
+
+  def _filter(self, session_groups):
+    return [sg for sg in session_groups if self._passes_all_filters(sg)]
+
+  def _passes_all_filters(self, session_group):
+    for f in self._filters:
+      if not f.passes(session_group):
+        return False
+    return True
+
+  def _sort(self, session_groups):
+    """Sorts 'session_groups' in place according to _request.col_params"""
+    def _create_key_func(extractor, none_is_largest):
+      """Returns a key_func to be used in list.sort() that sorts session groups
+      by the value extracted by extractor. None extracted values will either
+      be considered largest or smallest as specified by the "none_is_largest"
+      boolean parameter. """
+      if none_is_largest:
```

Comment:
FWIW, this could just return a single function that references none_is_largest, like

```python
def key_func(session_group):
  value = extractor.extract(session_group)
  none_column = value is None if none_is_largest else value is not None
  return (none_column, value)
```